lib/topological_inventory/providers/common/heartbeat.rb

```diff
@@ -0,0 +1,171 @@
+              sleep(HEARTBEAT_QUEUE_THREAD_TIMEOUT_STEP)
+
+              current_time_thread += HEARTBEAT_QUEUE_THREAD_TIMEOUT_STEP
+              if current_time_thread > self.max_thread_timeout
+                stop_queue_thread
+              else
+                self.queue_tick
+              end
+            end
+          end
+        end
+
+        def run_thread_queue_in_parallel_with
+          run_queue_thread_with_timeout
+
+          return_value = yield
+        ensure
+          stop_queue_thread
+
+          return_value
+        end
+
+        def run_thread_with_timeout
+          self.finished = Concurrent::AtomicBoolean.new(false)
+
+          Thread.new do
+            current_time_thread = 0
+            until finished.value
+              sleep(HEARTBEAT_THREAD_TIMEOUT_STEP)
+
+              current_time_thread += HEARTBEAT_THREAD_TIMEOUT_STEP
+              if current_time_thread > self.max_thread_timeout
+                stop
+              else
+                self.touch_heartbeat_file
+              end
+            end
+          end
+        end
+
+        def run_in_parallel_with
+          run_thread_with_timeout
+
+          return_value = yield
+        ensure
+          stop
+
+          return_value
+        end
+
+        def queue_tick
+          self.heartbeat_queue << :tick
+        end
+
+        def stop
+          self.finished = Concurrent::AtomicBoolean.new(true)
+        end
+
+        def stop_queue_thread
+          self.queue_thread_finished = Concurrent::AtomicBoolean.new(true)
+        end
+
+        def touch_heartbeat_file
+          File.write(self.heartbeat_file, (Time.now.utc + HEARTBEAT_CHECK_TIMEOUT).to_s)
+        end
+
+        def create_heartbeat_dir
+          full_heartbeat_dir = File.expand_path(heartbeat_dir, __FILE__)
+          Dir.mkdir(full_heartbeat_dir) unless File.exist?(full_heartbeat_dir)
+        end
+
+        def heartbeat_dir
+          self.class.heartbeat_dir
+        end
+
+        def heartbeat_file
+          self.class.heartbeat_file(self.name)
+        end
+
+        def self.heartbeat_dir
+          File.expand_path('tmp','.')
+        end
+
+        def self.heartbeat_file_path(name)
+          "#{heartbeat_dir}/heartbeat_file-#{name}.hb"
+        end
+
+        def self.heartbeat_file(name)
+          File.expand_path(self.heartbeat_file_path(name), __FILE__)
+        end
+
+        def self.check(name)
+          require 'time'
+
+          hb_file = heartbeat_file(name)
+
+          if File.exist?(hb_file)
+            current_time = Time.now.utc
+            contents     = File.read(hb_file)
+            mtime        = File.mtime(hb_file).utc
+
+            timeout      = if contents.empty?
+                             (mtime + HEARTBEAT_CHECK_TIMEOUT).utc
+                           else
+                             Time.parse(contents).utc
+                           end
+
+            current_time < timeout
```

Comment:
`<` is more confusing in specs than `<=` ... but in real usage it doesn't matter